File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.py
def solution(mats, park):
    answer = -1
    
    mats.sort(reverse=True) # 깔 수 있는 가장 큰 돗자리를 찾음
    
    opt_park = []
    for p in park:
        opt_p = []
        for l in p:
            if l != "-1": opt_p.append(0)
            else: opt_p.append(1)
        opt_park.append(opt_p)
    
    for m in mats:
        i = 0
        space = 0
        while i <= len(park)-m:
            for j in range(0, len(park[0])-m+2):
                if sum(opt_park[i][j:j+m]) == m:
                    for k in range(i, i+m):
                        temp = sum(opt_park[k][j:j+m])
                        if temp != m:
                            space = 0
                            break
                        space += temp
                    if space == m*m: return m
                else:
                    space = 0
            i += 1
    
    return answer

File: 프로그래머스/1/340198. ［PCCE 기출문제］ 10번 ／ 공원/［PCCE 기출문제］ 10번 ／ 공원.java
import java.util.*;

class Solution {
    int r, c;
    
    public int solution(int[] mats, String[][] park) {
        int answer = -1;
        r = park.length;
        c = park[0].length;
        
        Integer[] boxedMats = Arrays.stream(mats).boxed().toArray(Integer[]::new);
        Arrays.sort(boxedMats, Collections.reverseOrder());
        
        for (Integer m: boxedMats) {
            for (int i = 0; i < r-m+1; i++) {
                for (int j = 0; j < c-m+1; j++) {
                    if (canPlace(i, j, m, park)) return m;
                }
            }
        }
        
        return answer;
    }
    
    private boolean canPlace(int row, int column, int size, String[][] park) {
        if (row + size > r || column + size > c) return false;
        for (int i = row; i < row+size; i++) {
            for (int j = column; j < column+size; j++) {
                if (!park[i][j].equals("-1")) return false;
            }
        }
        return true;
    }
}